### ISSUE TITLE
`gem_package` fails because it is using `which` of `chef-sugar` if available

### ISSUE THREAD

**Issue by aried3r** [chef/chef#10512]
## Description

First of all, I don't know if this is the right place, but since a similar issue has been raised previously here (https://github.com/chef/chef/issues/9776), and `chef-sugar` doesn't seem to have an issue tracker, I'll post it here. If this is the wrong place, please let me know!

**TL;DR:** Using `gem_package` when `chef-sugar` is present, because `chef-sugar`'s `which` is being used instead of Chef's .

That being said, I am using the sous-chef nginx cookbook to install nginx with Passenger. The recipe to install contains the [following line](https://github.com/sous-chefs/nginx/blob/a7141cc9f9e9f8f69c663a1f578f99e2c3f114e1/resources/install.rb#L330):

```Ruby
gem_package 'rake' if install_rake?
```

In an older setup we also use nginx cookbook v9.0.0 which contains a [similar line](https://github.com/sous-chefs/nginx/blob/da2a01b006b9f42cc750877a63a9e4053e8a16dd/recipes/passenger.rb#L28). The output below will be from v9.0.0 since that has been sent to me.

```Ruby
gem_package 'rake' if node['nginx']['passenger']['install_rake']
```

Now, we also use the `firewall` cookbook in v2.7.0 which contains the [following line](https://github.com/chef-cookbooks/firewall/blob/56045ee7e109ce7287654aa360320e1599455bcd/recipes/default.rb#L20-L22)

```Ruby
chef_sugar_cookbook_version = Gem::Version.new(run_context.cookbook_collection['chef-sugar'].metadata.version)

include_recipe 'chef-sugar' if chef_sugar_cookbook_version < Gem::Version.new('4.0.0')
```

It is indeed the only cookbook in our setup that includes `chef-sugar`, we use v5.1.8 of `chef-sugar` right now, I've verified on our side that this check evaluates to `false`. So I'm not sure why it is being included or how `chef-sugar` is being used. `chef-sugar` includes its own implementation of `which`, which seems to be wrongly used in this case, instead of Chef's.

Chef's [`rubygems.rb`](https://github.com/chef/chef/blob/8c67bf7746301cb73dea14917a7a91f92a37eb13/lib/chef/provider/package/rubygems.rb#L439-L441):
```Ruby
def find_gem_by_path
  which("gem", extra_path: RbConfig::CONFIG["bindir"])
end
```

`chef-sugar`'s [`shell.rb`](https://github.com/chef/chef-sugar/blob/037f3ed21df513a1411b8c3939897cd78978deb4/lib/chef/sugar/shell.rb#L33-L46)
```Ruby
def which(cmd)
  if Pathname.new(cmd).absolute?
    File.executable?(cmd) ? cmd : nil
  else
    paths = ENV['PATH'].split(::File::PATH_SEPARATOR) + %w(/bin /usr/bin /sbin /usr/sbin)

    paths.each do |path|
      possible = File.join(path, cmd)
      return possible if File.executable?(possible)
    end

    nil
  end
end
```

A similar issue has been reported here https://github.com/chef/chef/issues/9776 and this same bug, I believe, has been reported in the comments here https://github.com/chef/chef/issues/9776#issuecomment-627419242 and fixed in the `apt` cookbook [here](https://github.com/chef-cookbooks/apt/commit/ab3fffc663e0459a6b59eccf6b47da44f06733cb#diff-241768b2e5fa81246e1a136579ec557e7e943b8773c5383fc9d5b1ac00b91948).

## Chef Version

Chef 16.6.14

## Platform Version

Ubuntu 18.04 LTS with chef-client 16.6.14.

## Replication Case

That's kind of hard, since we're updating a very old setup to be more modern, first verifying it works as intended. It uses librarian with knife-solo, both deprecated.

## Client Output

```
  * gem_package[rake] action install[2020-10-16T13:38:53+00:00] INFO: Processing gem_package[rake] action install (nginx::passenger line 28)


    ================================================================================
    Error executing action `install` on resource 'gem_package[rake]'
    ================================================================================

    ArgumentError
    -------------
    wrong number of arguments (given 2, expected 1)

    Resource Declaration:
    ---------------------
    # In /home/vagrant/chef-solo/local-mode-cache/cache/cookbooks/nginx/recipes/passenger.rb

     28: gem_package 'rake' if node['nginx']['passenger']['install_rake']
     29:

    Compiled Resource:
    ------------------
    # Declared in /home/vagrant/chef-solo/local-mode-cache/cache/cookbooks/nginx/recipes/passenger.rb:28:in `from_file'

    gem_package("rake") do
      package_name "rake"
      action [:install]
      default_guard_interpreter :default
      declared_type :gem_package
      cookbook_name "nginx"
      recipe_name "passenger"
    end

    System Info:
    ------------
    chef_version=16.6.14
    platform=ubuntu
    platform_version=18.04
    ruby=ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux]
    program_name=/usr/bin/chef-solo
    executable=/opt/chef/bin/chef-solo
```

## Stacktrace

```
>>>> Caused by ArgumentError: wrong number of arguments (given 2, expected 1)
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-sugar-5.1.11/lib/chef/sugar/shell.rb:132:in `which'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/provider/package/rubygems.rb:440:in `find_gem_by_path'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/provider/package/rubygems.rb:414:in `initialize'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource.rb:1507:in `new'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource.rb:1507:in `provider_for_action'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource.rb:599:in `block in run_action'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource.rb:626:in `with_umask'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource.rb:598:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/runner.rb:74:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/runner.rb:108:in `block in run_all_actions'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/runner.rb:108:in `each'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/runner.rb:108:in `run_all_actions'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/runner.rb:132:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource_collection/resource_list.rb:96:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource_collection/stepable_iterator.rb:54:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/resource_collection/resource_list.rb:94:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/2.7.0/forwardable.rb:235:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/runner.rb:130:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/client.rb:684:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/client.rb:679:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/client.rb:679:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/client.rb:703:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/client.rb:283:in `run'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/application.rb:305:in `run_with_graceful_exit_option'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/application.rb:281:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/local_mode.rb:42:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/application.rb:264:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/application/base.rb:337:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/application.rb:67:in `run'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-16.6.14/lib/chef/application/solo.rb:60:in `run'
/opt/chef/embedded/lib/ruby/gems/2.7.0/gems/chef-bin-16.6.14/bin/chef-solo:24:in `<top (required)>'
/usr/bin/chef-solo:155:in `load'
/usr/bin/chef-solo:155:in `<main>'
```

**Comment by tas50** [comment]
Potential workaround: https://github.com/chef/chef/pull/10628

**Comment by lamont-granquist** [comment]
The correct way to fix this looks like:

1.  backporting the chef-utils which into chef-sugar, including support for the optional arguments -- being careful to preserve the wiring to the default_paths.

2.  having chef-sugar disable the injection of its which mixin enirely on chef-client >= 16 so that it doesn't try to conflict any more

there doesn't appear to be any inherent conflict between the two implementations, but the chef-utils version has features and bugfixes that are lacking in the chef-sugar version.

**Comment by tas50** [comment]
This should be resolved in chef-sugar 5.1.12 with this merge: https://github.com/chef/chef-sugar/pull/21

You'll want to upgrade to the latest chef-sugar before trying to write code that runs on Chef Infra Client 16+ as well as older releases.
